### PR TITLE
Introduce a SecretSource implementation for .properties file

### PR DESCRIPTION
**Pull request by Smasherr**

### CODE REVIEW

File: plugin/pom.xml [hunk 1]
```diff
@@ -97,6 +97,18 @@
       <version>0.50.40</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>credentials</artifactId>
```
Comment (casz) [inline plugin/pom.xml]: Credential is already used in integration plugin, move the test to there? :)

**Comment by Smasherr** [comment]
Not sure if the default directory `/run/secrets` makes sense, since that's docker related.
Maybe there are also suggestions about better naming for the environment variable?

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public static final String SECRETS_DEFAULT_PATH = "/run/secrets/secrets.properties";
+
+    private final Properties secrets = new Properties();
+
+    public PropertiesSecretSource() {
+        final String secretsEnv = System.getenv("SECRETS");
+        final String secretsPath = secretsEnv == null ? SECRETS_DEFAULT_PATH : secretsEnv;
+        try (InputStream input = new FileInputStream(secretsPath)) {
+            secrets.load(input);
+        }
```
Comment (oleg-nenashev) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: We usually do not put line breaks after closing brackets

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * This {@link SecretSource} implementation allows to use a .properties file for providing secrets.
+ * The default file path is {@code /run/secrets/secrets.properties}, which can be changed via
+ * {@code SECRETS} environment variable.
+ *
+ * @author <a href="mailto:d.estermann.de@gmail.com">Daniel Estermann</a>
```
Comment (oleg-nenashev) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]:
```suggestion
 * @author <a href="mailto:d.estermann.de@gmail.com">Daniel Estermann</a>
 * @since TODO
```

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+ */
+@Extension
```
Comment (oleg-nenashev) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: If you do not want this class to be used as API, you could put `@Restricted(NoExternalUse.class)` here

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    private static final Logger LOGGER = Logger.getLogger(PropertiesSecretSource.class.getName());
+
+    /**
+     * Default path for .properties file
+     */
+    public static final String SECRETS_DEFAULT_PATH = "/run/secrets/secrets.properties";
+
+    private final Properties secrets = new Properties();
+
+    public PropertiesSecretSource() {
```
Comment (oleg-nenashev) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: Would be nice to run the retrieval upon first request instead of doing it through constructor. It would allow this code to be invoked on-demand, especially if the extension is disabled later

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    public Optional<String> reveal(String secret) {
+        // lazy initialization
+        if (secrets == null) {
+            secrets = new Properties();
+            final String secretsEnv = System.getenv("SECRETS");
+            final String secretsPath = secretsEnv == null ? SECRETS_DEFAULT_PATH : secretsEnv;
+            try (InputStream input = new FileInputStream(secretsPath)) {
+                secrets.load(input);
+            } catch (FileNotFoundException fnfe) {
+                LOGGER.log(Level.WARNING,
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: I don't like this being a warning as I'm not using this feature why should I care that the file doesn't exist?

I think one of two things should be done:
1. just check if the file exists before trying to load?
2. move the log level down to FINE or reword it to say this is normal and have it at INFO?

Unless I'm misunderstanding how this works

Comment (casz) [reply plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: Reveal secrets should have no side effects.

File: plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java [hunk 7]
```diff
@@ -41,14 +42,13 @@
...
-                    "Source properties file has not been found in the specified location: {0}",
-                    secretsPath);
-            } catch (IOException ioe) {
-                LOGGER.log(Level.WARNING, "Source properties file could not be loaded", ioe);
+            final File secretsFile = new File(secretsPath);
+            if (secretsFile.exists()) {
+                try (InputStream input = new FileInputStream(secretsFile)) {
+                    secrets.load(input);
+                } catch (IOException ioe) {
+                    LOGGER.log(Level.WARNING, "Source properties file could not be loaded", ioe);
```
Comment (timja) [inline plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/PropertiesSecretSource.java]: probably worth adding the file location to the error message?
